### ISSUE TITLE
[Jepsen] [YSQL] Inconsistency in counter test

### ISSUE THREAD

**Issue by frozenspider** [YugaByte/yugabyte-db#1532]
(Does not affect YCQL)
In an YSQL variant of Jepsen `counter` test (sends queries like `UPDATE t SET count = count + 1 WHERE id = 0`), we're seeing history like this:
```
2019-06-10 10:31:29,707{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:invoke	:add	1
2019-06-10 10:31:29,707{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:invoke	:add	1
2019-06-10 10:31:29,707{GMT}	INFO	[jepsen worker 3] jepsen.util: 3	:invoke	:add	1
2019-06-10 10:31:29,707{GMT}	INFO	[jepsen worker 2] jepsen.util: 2	:invoke	:add	1
2019-06-10 10:31:29,707{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:invoke	:add	1
2019-06-10 10:31:29,764{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:ok	:add	1
2019-06-10 10:31:29,818{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:ok	:add	1
2019-06-10 10:31:29,824{GMT}	INFO	[jepsen worker 3] jepsen.util: 3	:ok	:add	1
2019-06-10 10:31:29,865{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:invoke	:add	1
2019-06-10 10:31:29,876{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:ok	:add	1
2019-06-10 10:31:29,920{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:invoke	:add	1
2019-06-10 10:31:29,925{GMT}	INFO	[jepsen worker 3] jepsen.util: 3	:invoke	:add	1
2019-06-10 10:31:29,933{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:ok	:add	1
2019-06-10 10:31:29,939{GMT}	INFO	[jepsen worker 3] jepsen.util: 3	:ok	:add	1
2019-06-10 10:31:29,977{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:invoke	:add	1
2019-06-10 10:31:29,990{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:ok	:add	1
2019-06-10 10:31:30,035{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:invoke	:add	1
2019-06-10 10:31:30,040{GMT}	INFO	[jepsen worker 3] jepsen.util: 3	:invoke	:add	1
2019-06-10 10:31:30,048{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:ok	:add	1
2019-06-10 10:31:30,092{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:invoke	:add	1
2019-06-10 10:31:30,106{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:ok	:add	1
2019-06-10 10:31:30,149{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:invoke	:add	1
2019-06-10 10:31:30,165{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:ok	:add	1
2019-06-10 10:31:30,207{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:invoke	:add	1
2019-06-10 10:31:30,218{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:ok	:add	1
2019-06-10 10:31:30,266{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:invoke	:add	1
2019-06-10 10:31:30,286{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:ok	:add	1
2019-06-10 10:31:30,319{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:invoke	:add	1
2019-06-10 10:31:30,331{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:ok	:add	1
2019-06-10 10:31:30,387{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:invoke	:add	1
2019-06-10 10:31:30,432{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:invoke	:add	1
2019-06-10 10:31:30,444{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:ok	:add	1
2019-06-10 10:31:30,544{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:invoke	:read	nil
2019-06-10 10:31:30,550{GMT}	INFO	[jepsen worker 1] jepsen.util: 1	:ok	:read	12
```
i.e. after 14 `ok` responses read request yield 12
Full logs: [logs-jepsen_2019-06-10_15-33-40.tar.gz](https://github.com/YugaByte/yugabyte-db/files/3275463/logs-jepsen_2019-06-10_15-33-40.tar.gz)
Reproducible with latest master (6d65f74394715911558aea5d4a1777d1ba5601f0), related to #938

**Comment by ndeodhar** [comment]
@frozenspider Is this for serializable or snapshot isolation?

**Comment by frozenspider** [comment]
@ndeodhar serializable, haven't tested with snapshot

**Comment by frozenspider** [comment]
Update: this test does not use explicit transaction, so it was actually using snapshot isolation default for single statements.
However, it's also reproducible with explicit transactions on both isolation levels.

**Comment by mbautin** [comment]
@georgeklees: take a look at pg_libpq-test.cc as an example.

**Comment by bmatican** [comment]
@georgeklees similarly, we have this `src/yb/client/serializable-txn-test.cc`, and the increment tests:
```
269 TEST_F(SerializableTxnTest, Increment) {
270   TestIncrements(true /* transactional */);
271 }
272
273 TEST_F(SerializableTxnTest, IncrementNonTransactional) {
274   TestIncrements(false /* transactional */);
275 }
```

**Comment by georgeklees** [comment]
With my commit for #1685 we now have C++ versions of these counter tests, all of which now pass even though they had been failing before @mbautin and @spolitov's diffs fixed things.

**Comment by frozenspider** [comment]
Resolved in 76fcf866d9fad634f563ef76a31f0b06b6f8b67c by @spolitov (for SERIALIZABLE) and 5dd4385ebc0af412f45e449ca4f27bfe930d370a by @mbautin (for REPEATABLE READ)
